File: new/pa.py
from jinja2 import Template
import os


BUILD_DIR="./build"

function_template = Template("""
void paged_attention({{out_dtype}}* out,
                    void* workspace_buffer,
                    {{dtype}}* query_ptr,
                    {{kv_dtype}}* key_cache_ptr,
                    {{kv_dtype}}* value_cache_ptr,
                    float scale,
                    const int num_seqs,
                    const int q_stride,
                    const int kv_block_stride,
                    const int kv_head_stride,
                    const int kv_seq_stride,
                    int* kv_indptr_ptr,
                    int* kv_page_indices_ptr,
                    int* kv_last_page_lens_ptr,
                    const float* alibi_slopes_ptr,
                    float logits_soft_cap,
                    const float* k_scale_ptr,
                    const float* v_scale_ptr,
                    const float* fp8_out_scale_ptr,
                    hipStream_t stream)
""")

header_template = Template("""
#include <hip/hip_bf16.h>
#include <hip/hip_runtime.h> 
#include "hip_compat.h"

#include <algorithm>
#include "dtype_fp8.cuh"
#include "quant_utils.cuh"
                           
                    
template <typename scalar_t,
          typename cache_t,
          vllm::Fp8KVCacheDataType KV_DTYPE,
          typename OUTT,
          int BLOCK_SIZE,
          int HEAD_SIZE,
          int NUM_THREADS,
          bool ALIBI_ENABLED,
          bool LOGITS_SOFT_CAP_ENABLED,
          int GQA_RATIO>
__global__ __launch_bounds__(NUM_THREADS) void paged_attention_ll4mi_QKV_mfma16_kernel(
    const scalar_t* __restrict__ q,      // [num_seqs, num_heads, head_size]
    const cache_t* __restrict__ k_cache, // [num_blocks, num_kv_heads,
                                         // head_size/x, block_size, x]
    const cache_t* __restrict__ v_cache, // [num_blocks, num_kv_heads,
                                         // head_size, block_size]
    const float scale,
    const int* __restrict__ kv_indptr,         // [num_seqs + 1]
    const int* __restrict__ kv_page_indices,   // [max_num_blocks]
    const int* __restrict__ kv_last_page_lens, // [num_seqs]
    const float* __restrict__ alibi_slopes,    // [num_heads]
    const int q_stride,
    const int kv_block_stride,
    const int kv_head_stride,
    const int kv_seq_stride,
    float* __restrict__ exp_sums,   // [num_seqs, num_heads, max_num_partitions]
    float* __restrict__ max_logits, // [num_seqs, num_heads,
                                    // max_num_partitions]
    scalar_t* __restrict__ out,     // [num_seqs, num_heads, max_num_partitions,
                                    // head_size]
    OUTT* __restrict__ final_out,   // [num_seqs, num_heads, head_size]
    float logits_soft_cap,
    const float* k_scale_ptr,
    const float* v_scale_ptr,
    const float* __restrict__ fp8_out_scale_ptr);
                           
template <typename scalar_t,
          typename OUTT,
          int HEAD_SIZE,
          int NUM_THREADS,
          int PARTITION_SIZE,
          int NPAR_LOOPS, bool ENABLE_LAST_PAGE_LENS>
__global__ __launch_bounds__(NUM_THREADS) void paged_attention_ll4mi_reduce_kernel(
    OUTT* __restrict__ out,                    // [num_seqs, num_heads, head_size]
    const float* __restrict__ exp_sums,        // [num_seqs, num_heads,
                                               // max_num_partitions]
    const float* __restrict__ max_logits,      // [num_seqs, num_heads,
                                               // max_num_partitions]
    const scalar_t* __restrict__ tmp_out,      // [num_seqs, num_heads,
                                               // max_num_partitions, head_size]
    const int* __restrict__ kv_indptr,         // [num_seqs + 1]
    const int* __restrict__ kv_last_page_lens, // [num_seqs]
    const int block_size,
    const int max_num_partitions,
    const float* __restrict__ fp8_out_scale_ptr);

{{function_define}};
""")


src_template = Template("""
#include <pa.h>

#define DIVIDE_ROUND_UP(a, b) (((a) + (b)-1) / (b))

{{function_define}}
{
    constexpr int num_kv_heads = {{num_kv_heads}};
    constexpr int num_heads       = {{num_heads}};
    constexpr int head_size       = {{head_size}};
    constexpr int max_num_partitions = {{max_num_partitions}};
    constexpr int PARTITION_SIZE = 256;
    constexpr int gqa_ratio = num_heads / num_kv_heads;
    assert(num_heads % num_kv_heads == 0);

    float* exp_sums_ptr   = reinterpret_cast<float*>(workspace_buffer.data_ptr());
    float* max_logits_ptr = exp_sums_ptr + (num_seqs * num_heads * max_num_partitions);
    {{dtype}}* tmp_out_ptr =
        reinterpret_cast<T*>(max_logits_ptr + (num_seqs * num_heads * max_num_partitions));

    constexpr int NTHR = 256;
    dim3 grid(num_seqs, max_num_partitions, num_kv_heads);
    dim3 block(NTHR);
    if(logits_soft_cap>0){
        paged_attention_ll4mi_QKV_mfma16_kernel<{{dtype}},                       
                                            {{kv_dtype}},                            {{fp8_kv_dtype}},                
                                            {{out_dtype}},                    
                                            {{block_size}},              
                                            head_size,               
                                            NTHR,                    
                                            {{alibi_enabled}},           
                                            true, 
                                            gqa_ratio>               
        <<<grid, block, 0, stream>>>(query_ptr,                      
                                     key_cache_ptr,                  
                                     value_cache_ptr,                
                                     scale,                          
                                     kv_indptr_ptr,                  
                                     kv_page_indices_ptr,            
                                     kv_last_page_lens_ptr,          
                                     alibi_slopes_ptr,               
                                     q_stride,                       
                                     kv_block_stride,                
                                     kv_head_stride,                 
                                     kv_seq_stride,                  
                                     exp_sums_ptr,                   
                                     max_logits_ptr,                 
                                     tmp_out_ptr,                    
                                     out_ptr,                        
                                     logits_soft_cap,                
                                     k_scale_ptr,                    
                                     v_scale_ptr,                    
                                     fp8_out_scale_ptr);
    }else{
       paged_attention_ll4mi_QKV_mfma16_kernel<{{dtype}},                       
                                            {{kv_dtype}},                                                        {{fp8_kv_dtype}},                
                                            {{out_dtype}},                    
                                            {{block_size}},              
                                            head_size,               
                                            NTHR,                    
                                            {{alibi_enabled}},           
                                            false, 
                                            gqa_ratio>               
        <<<grid, block, 0, stream>>>(query_ptr,                      
                                     key_cache_ptr,                  
                                     value_cache_ptr,                
                                     scale,                          
                                     kv_indptr_ptr,                  
                                     kv_page_indices_ptr,            
                                     kv_last_page_lens_ptr,          
                                     alibi_slopes_ptr,               
                                     q_stride,                       
                                     kv_block_stride,                
                                     kv_head_stride,                 
                                     kv_seq_stride,                  
                                     exp_sums_ptr,                   
                                     max_logits_ptr,                 
                                     tmp_out_ptr,                    
                                     out_ptr,                        
                                     logits_soft_cap,                
                                     k_scale_ptr,                    
                                     v_scale_ptr,                    
                                     fp8_out_scale_ptr);
    }
 
    dim3 reduce_grid(num_heads, num_seqs);
    dim3 reduce_block(head_size);
    constexpr int npar_loops = DIVIDE_ROUND_UP(max_num_partitions, warpSize);
    paged_attention_ll4mi_reduce_kernel<{{dtype}}, {{out_dtype}}, head_size, head_size, PARTITION_SIZE, npar_loops, {{enable_last_page_lens}}> 
    <<<reduce_grid, reduce_block, 0, stream>>>(out_ptr,                                        
                                                exp_sums_ptr,        
                                                max_logits_ptr,                                 
                                                tmp_out_ptr,                                   
                                                kv_indptr_ptr,                                 
                                                kv_last_page_lens_ptr,                         
                                                {{block_size}},                                    
                                                max_num_partitions,                            
                                                fp8_out_scale_ptr);
}
""")


makefile_template = Template("""
CXX = hipcc
CXXFLAGS = -fPIC -O3 -std=c++17
INCLUDES = -Iater/csrc/include

TARGET = libpa.so
SRCS = pa.cpp pa.cu
OBJS = $(SRCS:.cpp=.o)

.PHONY: all clean

all: $(TARGET)

$(TARGET): $(OBJS)
	$(CXX) -shared $(OBJS) -o $@

%.o: %.cpp
	$(CXX) $(CXXFLAGS) $(INCLUDES) -c $< -o $@

clean:
	rm -f $(OBJS) $(TARGET)
""")

def init_build_dir():
    if not os.path.exists(BUILD_DIR):
        os.makedirs(BUILD_DIR)
    os.system(f"cp ./pa.cu {BUILD_DIR}/")

def compile(num_kv_heads, num_seqs, num_heads, head_size, max_num_partitions, dtype, kv_dtype, fp8_kv_dtype, out_dtype, block_size, alibi_enabled):
    # pass
    init_build_dir()
    header_file = header_template.render(function_define=function_template.render(out_dtype=out_dtype, dtype=dtype, kv_dtype=kv_dtype))
    # print(header_file)
    with open(f"{BUILD_DIR}/pa.h", "w") as f:
        f.write(header_file)
    src_file = src_template.render(num_kv_heads=num_kv_heads, num_seqs=num_seqs, num_heads=num_heads, head_size=head_size, max_num_partitions=max_num_partitions, dtype=dtype, kv_dtype=kv_dtype, fp8_kv_dtype=fp8_kv_dtype, out_dtype=out_dtype, block_size=block_size, alibi_enabled=alibi_enabled, enable_last_page_lens="true" if block_size > 1 else "false", function_define=function_template.render(out_dtype=out_dtype, dtype=dtype, kv_dtype=kv_dtype))
    with open(f"{BUILD_DIR}/pa.cpp", "w") as f:
        f.write(src_file)

    makefile_file = makefile_template.render()
    with open(f"{BUILD_DIR}/Makefile", "w") as f:
        f.write(makefile_file)

if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--num_kv_heads", type=int, required=True)
    parser.add_argument("--num_seqs", type=int, required=True)
    parser.add_argument("--num_heads", type=int, required=True)
    parser.add_argument("--head_size", type=int, required=True)
    parser.add_argument("--max_num_partitions", type=int, required=True)
    parser.add_argument("--dtype", type=str, required=True)
    parser.add_argument("--kv_dtype", type=str, required=True)
    parser.add_argument("--fp8_kv_dtype", type=str, required=True)
    parser.add_argument("--out_dtype", type=str, required=True)
    parser.add_argument("--block_size", type=int, required=True)
    parser.add_argument("--alibi_enabled", type=str, required=True)
    args = parser.parse_args()
    compile(args.num_kv_heads, args.num_seqs, args.num_heads, args.head_size, args.max_num_partitions, args.dtype, args.kv_dtype, args.fp8_kv_dtype, args.out_dtype, args.block_size, args.alibi_enabled)

File: cpp/pa.cpp
#include "utils.h"
#include <fmt/core.h>
#include "pa.h"


void paged_attention_ragged(
    torch::Tensor& out, // [num_seqs, num_heads, head_size]
    torch::Tensor& workspace_buffer,
    torch::Tensor& query,       // [num_seqs, num_heads, head_size]
    torch::Tensor& key_cache,   // [num_blocks, num_heads, block_size, head_size] or
                                // [num_blocks, block_size, num_heads, head_size]
    torch::Tensor& value_cache, // [num_blocks, num_heads, block_size, head_size] or
                                // [num_blocks, block_size, num_heads, head_size]
    double scale,
    torch::Tensor& kv_indptr,         // [num_seqs + 1]
    torch::Tensor& kv_page_indices,   // [max_num_blocks]
    std::optional<torch::Tensor>& kv_last_page_lens, // [num_seqs]
    int64_t block_size,
    int64_t max_num_partitions,
    const std::optional<torch::Tensor>& alibi_slopes,
    const std::string& kv_cache_dtype,
    const std::string& kv_cache_layout,
    float logits_soft_cap,
    torch::Tensor& k_scale,
    torch::Tensor& v_scale,
    const std::optional<torch::Tensor>& fp8_out_scale){
    const void* stream = reinterpret_cast<void*>(at::hip::getCurrentHIPStream().stream());
    std::string dtype;
    std::string kv_dtype;
    if(kv_cache_dtype == "auto")
    {
        if(query.dtype() == at::ScalarType::Half)
        {
            dtype = "_Float16";
            kv_dtype = "_Float16";
        }
        else if(query.dtype() == at::ScalarType::BFloat16)
        {
            dtype = "__hip_bfloat16";
            kv_dtype = "__hip_bfloat16";
        }
        else
        {
            TORCH_CHECK(false, "Unsupported data type: ", query.dtype());
        }
    }
    else if(kv_cache_dtype == "fp8" || kv_cache_dtype == "fp8_e4m3")
    {
        if(query.dtype() == at::ScalarType::Half)
        {
            dtype = "_Float16";
            kv_dtype = "uint8_t";
        }
        else if(query.dtype() == at::ScalarType::BFloat16)
        {
            dtype = "__hip_bfloat16";
            kv_dtype = "uint8_t";
        }
        else
        {
            TORCH_CHECK(false, "Unsupported data type: ", query.dtype());
        }
    }

    const int num_kv_heads = kv_cache_layout=="HND" ? key_cache.size(1) : key_cache.size(2);
    int num_seqs        = query.size(0);
    int num_heads       = query.size(1);
    int head_size       = query.size(2);
    int q_stride        = query.stride(0);
    int kv_block_stride = key_cache.stride(0);
    int kv_head_stride  = kv_cache_layout == "HND" ? key_cache.stride(1) : key_cache.stride(2);
    int kv_seq_stride   = kv_cache_layout == "HND" ? key_cache.stride(2) : key_cache.stride(1);
    std::string cmd = fmt::format(R"(python3 pa.py --num_kv_heads={num_kv_heads} \
                --num_seqs={num_seqs} \
                --num_heads={num_heads} \
                --head_size={head_size} \
                --max_num_partitions={max_num_partitions} \
                --dtype={dtype} \
                --kv_dtype={kv_dtype} \
                --fp8_kv_dtype={fp8_kv_dtype} \
                --out_dtype={out_dtype} \
                --block_size={block_size} \
                --alibi_enabled={alibi_enabled} \
                --enable_last_page_lens={enable_last_page_lens})",
                fmt::arg("num_kv_heads", num_kv_heads),
                fmt::arg("num_seqs", num_seqs),
                fmt::arg("num_heads", num_heads),
                fmt::arg("head_size", head_size),
                fmt::arg("max_num_partitions", max_num_partitions),
                fmt::arg("dtype", dtype),
                fmt::arg("kv_dtype", kv_dtype),
                fmt::arg("fp8_kv_dtype", kv_cache_dtype), 
                fmt::arg("out_dtype", dtype),
                fmt::arg("block_size", block_size),
                fmt::arg("alibi_enabled", alibi_slopes ? "true" : "false"),
                fmt::arg("enable_last_page_lens", block_size > 1 ? "true" : "false"));
    executeCmd(cmd);
    void* query_ptr = query.data_ptr();
    void* key_cache_ptr = key_cache.data_ptr();
    void* value_cache_ptr = value_cache.data_ptr();
    void* workspace_buffer_ptr = workspace_buffer.data_ptr();
    int* kv_indptr_ptr         = kv_indptr.data_ptr<int>();
    int* kv_page_indices_ptr   = kv_page_indices.data_ptr<int>();
    int* kv_last_page_lens_ptr = kv_last_page_lens ? kv_last_page_lens.value().data_ptr<int>() : nullptr;
    const float* k_scale_ptr = k_scale.data_ptr<float>();
    const float* v_scale_ptr = v_scale.data_ptr<float>();
    const float* fp8_out_scale_ptr =
        fp8_out_scale ? fp8_out_scale.value().data_ptr<float>() : nullptr;
    void* out_ptr = out.data_ptr();
    const float* alibi_slopes_ptr = alibi_slopes ? alibi_slopes.value().data_ptr<float>() : nullptr;
    std::string folder = fmt::format("pa_ragged_{num_kv_heads}_{num_seqs}_{num_heads}_{head_size}_{max_num_partitions}_{dtype}_{kv_dtype}_{fp8_kv_dtype}_{out_dtype}_{block_size}_{alibi_enabled}_{enable_last_page_lens}",
                fmt::arg("num_kv_heads", num_kv_heads),
                fmt::arg("num_seqs", num_seqs),
                fmt::arg("num_heads", num_heads),
                fmt::arg("head_size", head_size),
                fmt::arg("max_num_partitions", max_num_partitions),
                fmt::arg("dtype", dtype),
                fmt::arg("kv_dtype", kv_dtype),
                fmt::arg("fp8_kv_dtype", kv_cache_dtype), 
                fmt::arg("out_dtype", dtype),
                fmt::arg("block_size", block_size),
                fmt::arg("alibi_enabled", alibi_slopes ? "true" : "false"),
                fmt::arg("enable_last_page_lens", block_size > 1 ? "true" : "false"));
    run_lib(folder, out_ptr, workspace_buffer_ptr, query_ptr, key_cache_ptr, value_cache_ptr, scale, num_seqs, q_stride, kv_block_stride, kv_head_stride, kv_seq_stride, kv_indptr_ptr, kv_page_indices_ptr, kv_last_page_lens_ptr, alibi_slopes_ptr, logits_soft_cap, k_scale_ptr, v_scale_ptr, fp8_out_scale_ptr, stream);
}